sklearn/preprocessing/_data.py

```diff
@@ -3236,14 +3236,21 @@ def _yeo_johnson_optimize(self, x):
 
         Like for Box-Cox, MLE is done via the brent optimizer.
         """
+        x_tiny = np.finfo(x.dtype).tiny
 
         def _neg_log_likelihood(lmbda):
             """Return the negative log likelihood of the observed data x as a
             function of lambda."""
             x_trans = self._yeo_johnson_transform(x, lmbda)
             n_samples = x.shape[0]
+            x_trans_var = x_trans.var()
 
-            loglike = -n_samples / 2 * np.log(x_trans.var())
+            # Reject transformed data that is constant
+            if x_trans_var < x_tiny:
+                return np.inf
```

Comment:
tiny is really small and will probably not catch what should be considered constant data in some cases.
What do you think about using ``_is_constant_feature`` (maybe changing the name) which is designed for that ?

Reply:
I updated the comment. This is more for detecting the runtime warning in `np.log` in the line below. As long as the `np.log(variance)` can be computed the likelihood can be computed as well.

It turns out that `np.log` can handle values below `tiny` as well:

```python
import numpy as np

np.log(np.finfo(np.float64).tiny * 1e-15)
# -742.83
```

`np.log` even works down to the smallest subnormal: 

```python
import numpy as np

finfo = np.finfo(np.float64)
finfo.smallest_subnormal
# 5e-324

np.log(finfo.smallest_subnormal)
# -744.44

np.log(finfo.smallest_subnormal * 0.5)
# -inf
```

This `x_trans_var < x_tiny` is used because of a valid threading concern raised here: https://github.com/scikit-learn/scikit-learn/pull/20653#discussion_r762636355 Originally, I caught the runtime warning, but [catch_warning](https://docs.python.org/3/library/warnings.html#warnings.catch_warnings) is not thread-safe.

Reply:
> I updated the comment. This is more for detecting the runtime warning in np.log in the line below. As long as the np.log(variance) can be computed the likelihood can be computed as well.

What I meant is that even if computable, its value would be meaningless. The variance is so small that it lies within the theoretical error bounds, meaning it's undistinguishable from a zero variance.

However, this situation should not appear very often, and even if it does, this lambda would not be the argmin anyway (unless all lambdas lead to constant x), so I'm so I'm ok with the ``tiny`` solution as well.